dv/uvm/core_ibex/Makefile

```diff
@@ -114,45 +211,123 @@ RISCV_DV_OPTS=--custom_target=${DV_DIR}/riscv_dv_extension \
+# the 'gen' stage for more verbose explanations of how this works.
+iss-var-deps := ISS ISS_OPTS
+-include $(OUT-SEED)/instr_gen/.iss-vars.mk
+iss-vars-prereq = $(if $(call vars-match,$(iss-var-deps),iss),,FORCE)
+
+$(OUT-SEED)/instr_gen/.iss.stamp: \
+  $(iss-vars-prereq) $(OUT-SEED)/instr_gen/.compile.stamp
+	@python3 ${GEN_DIR}/run.py \
+     --o=$(OUT-SEED)/instr_gen ${GEN_OPTS} \
      --steps=iss_sim \
      ${COMMON_OPTS} \
      --iss="${ISS}" \
      --iss_opts="${ISS_OPTS}" \
-     ${RISCV_DV_OPTS} \
+     ${RISCV_DV_OPTS}
+	$(call dump-vars,$(iss-var-deps),$(@D)/.iss-vars.mk,iss)
+	@touch $@
+
+.PHONY: iss_sim
+iss_sim: $(OUT-SEED)/instr_gen/.iss.stamp
+
 
+###############################################################################
 # Compile ibex core TB
-compile:
-	mkdir -p ${OUT}/rtl_sim;
-	python3 ./sim.py \
+#
+# Note that (unlike everything else) this doesn't depend on the seed: the DUT
+# doesn't depend on which test we're running!
+#
+# It does, however, depend on various variables. These are listed in
+# compile-var-deps. See the 'gen' stage for more verbose explanations of how
+# the variable dumping works.
+#
+# The compiled ibex testbench (obviously!) also depends on the design and the
+# DV code. The clever way of doing this would be to look at a dependency
+# listing generated by the simulator as a side-effect of doing the compile (a
+# bit like using the -M flags with a C compiler). Unfortunately, that doesn't
+# look like it's particularly easy, so we'll just depend on every .v, .sv or
+# .svh file in the dv or rtl directories. Note that this variable is set with
+# '=', rather than ':='. This means that we don't bother running the find
+# commands unless we need the compiled testbench.
+all-verilog = \
+  $(shell find ../../../rtl -name '*.v' -o -name '*.sv' -o -name '*.svh') \
+  $(shell find ../.. -name '*.v' -o -name '*.sv' -o -name '*.svh')
+
+compile-var-deps := COMMON_OPTS SIMULATOR COV WAVES COMPILE_OPTS
+-include $(OUT)/.compile-vars.mk
+compile-vars-prereq = $(if $(call vars-match,$(compile-var-deps),comp),,FORCE)
+
+$(OUT)/rtl_sim/.compile.stamp: \
+  $(compile-vars-prereq) $(all-verilog) $(risc-dv-files) | $(OUT)
+	@python3 ./sim.py \
      --o=${OUT} \
      --riscv_dv_root=${GEN_DIR} \
      --steps=compile \
      ${COMMON_OPTS} \
      --simulator="${SIMULATOR}" \
      --en_cov=${COV} \
      --en_wave=${WAVES} \
-     --cmp_opts="${COMPILE_OPTS}" \
+     --cmp_opts="${COMPILE_OPTS}"
+	$(call dump-vars,$(compile-var-deps),$(OUT)/.compile-vars.mk,comp)
+	@touch $@
 
-# Run ibex RTL simulation with random instructions
-rtl_sim:
-	mkdir -p ${OUT}/rtl_sim
-	python3 ./sim.py \
-     --o=${OUT} \
+.PHONY: compile
+compile: $(OUT)/rtl_sim/.compile.stamp
+
+###############################################################################
+# Run ibex RTL simulation with generated programs
+#
+# Because we compile a TB once rather than for each seed, we have to copy in
```

Comment:
One question based off of some messing around I did with the new Makefile format.
After running `make ITERATIONS=1 TEST=riscv_rand_instr_test SEED=123 instr`, I ran `make ITERAITONS=1 TEST=riscv_rand_instr_test compile`, which then placed the output vcs binaries into `out/rtl_sim`.
Then, I ran `make ITERATIONS=1 TEST=riscv_rand_instr_test SEED=123 sim`, which then recompiled the entire testbench into `out/seed-123/rtl_sim`, instead of just copying the `out/rtl_sim` directory over, which is what this sections seems to be suggesting.
Is the full recompilation in this situation intended?

Reply:
The full recompilation definitely isn't intended! Oops! I'll have a look at that this morning.

About the `.metadata` directory suggestion, where exactly do you think that should go? At the moment, we're creating:
  - out/seed-123/instr_gen/.gen.stamp
  - out/seed-123/instr_gen/.compile.stamp
  - out/seed-123/instr_gen/.iss.stamp
  - out/rtl_sim/.compile.stamp
  - out/seed-123/rtl_sim/.compile.stamp
  - out/seed-123/rtl_sim/.run.stamp

This is 3 different directories (dictated by test flags, the design, and "test flags & design", respectively). Are you proposing a parallel directory tree at out/.metadata? That seems like it would be more confusing than helpful. Or have I misunderstood?

Reply:
Right, I think the recompilation is fixed. I'll push up a patch with the fix in a second. I've actually made some more changes which make what's going on a bit clearer for the user: the checks for whether variables have changed now print out a message if a variable differs.

The thing that caused the recompilation is that compiling the TB depended on `COMMON_OPTS` which contains the test list (which you, quite reasonably, weren't setting in your command lines above). The new patch now distinguishes between `COMMON_OPTS` and `TEST_OPTS`. The latter is used for targets that depend on which tests we're running.

Reply:
Sorry, I don't think my suggestion was very clear, let me try to elaborate a bit.
I think the out/rtl_sim/.compile.stamp file is fine as-is, since only that one metadata file is generated in a 'non-seed' directory.
As for all of the .stamp files generated in every seed-### directory, my suggestion would be create a directory `out/seed-123/.metadata`, and tag the files as follows:
* out/seed-123/.metadata/.instr_gen.gen.stamp
* out/seed-123/.metadata/.instr_gen.compile.stamp
* out/seed-123/.metadata/.instr_gen.iss.stamp
* out/seed-123/.metadata/.rtl_sim.compile.stamp
* out/seed-123/.metadata/.rtl_sim.run.stamp